boa_engine/src/syntax/parser/expression/identifiers.rs

```diff
@@ -0,0 +1,238 @@
+//! Identifiers parsing.
+//!
+//! More information:
+//!  - [ECMAScript specification][spec]
+//!
+//! [spec]: https://tc39.es/ecma262/#sec-identifiers
+//!
+
+use crate::syntax::{
+    ast::{node::Identifier, Keyword},
+    lexer::{Error as LexError, TokenKind},
+    parser::{cursor::Cursor, AllowAwait, AllowYield, ParseError, TokenParser},
+};
+use boa_interner::{Interner, Sym};
+use boa_profiler::Profiler;
+use std::io::Read;
+
+const RESERVED_IDENTIFIERS_STRICT: [Sym; 9] = [
+    Sym::IMPLEMENTS,
+    Sym::INTERFACE,
+    Sym::LET,
+    Sym::PACKAGE,
+    Sym::PRIVATE,
+    Sym::PROTECTED,
+    Sym::PUBLIC,
+    Sym::STATIC,
+    Sym::YIELD,
+];
+
+/// Identifier reference parsing.
+///
+/// More information:
+///  - [ECMAScript specification][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#prod-IdentifierReference
+#[derive(Debug, Clone, Copy)]
+pub(in crate::syntax::parser) struct IdentifierReference {
+    allow_yield: AllowYield,
+    allow_await: AllowAwait,
+}
+
+impl IdentifierReference {
+    /// Creates a new `IdentifierReference` parser.
+    pub(in crate::syntax::parser) fn new<Y, A>(allow_yield: Y, allow_await: A) -> Self
+    where
+        Y: Into<AllowYield>,
+        A: Into<AllowAwait>,
+    {
+        Self {
+            allow_yield: allow_yield.into(),
+            allow_await: allow_await.into(),
+        }
+    }
+}
+
+impl<R> TokenParser<R> for IdentifierReference
+where
+    R: Read,
+{
+    type Output = Identifier;
+
+    fn parse(
+        self,
+        cursor: &mut Cursor<R>,
+        interner: &mut Interner,
+    ) -> Result<Self::Output, ParseError> {
+        let _timer = Profiler::global().start_event("IdentifierReference", "Parsing");
+
+        let token = cursor.next(interner)?.ok_or(ParseError::AbruptEnd)?;
+
+        match token.kind() {
+            TokenKind::Identifier(ident)
+                if cursor.strict_mode() && RESERVED_IDENTIFIERS_STRICT.contains(ident) =>
```

Comment:
These `contains()` calls can in the future be exchanged for a phf lookup, which is O(1), since it's calculated in compile time.